pkg/models/commandvalue.go

```diff
@@ -30,632 +26,593 @@ const (
 	Type string
-	// NumericValue is a byte slice with a maximum capacity of
-	// 64 bytes, used to hold a numeric value returned by a
-	// ProtocolDriver instance. The value can be converted to
-	// its native type by referring to the the value of ResType.
-	NumericValue []byte
-	// stringValue is a string value returned as a value by a ProtocolDriver instance.
-	stringValue string
-	// BinValue is a binary value with a maximum capacity of 16 MB,
-	// used to hold binary values returned by a ProtocolDriver instance.
-	BinValue []byte
+	// Value holds value returned by a ProtocolDriver instance.
+	// The value can be converted to its native type by referring to ValueType.
+	Value interface{}
 }
 
-// NewBoolValue creates a CommandValue of Type Bool with the given value.
-func NewBoolValue(DeviceResourceName string, origin int64, value bool) (cv *CommandValue, err error) {
-	cv = &CommandValue{DeviceResourceName: DeviceResourceName, Origin: origin, Type: v2.ValueTypeBool}
-	err = encodeValue(cv, value)
-	return
-}
-
-// NewBoolArrayValue creates a CommandValue of Type BoolArray with the given value.
-func NewBoolArrayValue(DeviceResourceName string, origin int64, value []bool) (cv *CommandValue, err error) {
-	jsonValue, err := json.Marshal(value)
-	cv = &CommandValue{
-		DeviceResourceName: DeviceResourceName,
-		Origin:             origin,
-		Type:               v2.ValueTypeBoolArray,
-		stringValue:        string(jsonValue),
+// NewCommandValue create a CommandValue according to the valueType supplied.
+func NewCommandValue(deviceResourceName string, valueType string, value interface{}) (*CommandValue, errors.EdgeX) {
+	err := validate(valueType, value)
+	if err != nil {
+		return nil, errors.NewCommonEdgeX(errors.KindServerError, "failed to create CommandValue", err)
 	}
 
-	return
+	return &CommandValue{
+		DeviceResourceName: deviceResourceName,
+		Type:               valueType,
+		Value:              value}, nil
 }
 
 // NewStringValue creates a CommandValue of Type string with the given value.
-func NewStringValue(DeviceResourceName string, origin int64, value string) (cv *CommandValue) {
-	cv = &CommandValue{DeviceResourceName: DeviceResourceName, Origin: origin, Type: v2.ValueTypeString, stringValue: value}
-	return
+func NewStringValue(deviceResourceName string, value string) (*CommandValue, errors.EdgeX) {
+	return NewCommandValue(deviceResourceName, v2.ValueTypeString, value)
+}
+
+// NewBoolValue creates a CommandValue of Type Bool with the given value.
+func NewBoolValue(deviceResourceName string, value bool) (*CommandValue, errors.EdgeX) {
+	return NewCommandValue(deviceResourceName, v2.ValueTypeBool, value)
+}
+
+// NewBoolArrayValue creates a CommandValue of Type BoolArray with the given value.
+func NewBoolArrayValue(deviceResourceName string, value []bool) (*CommandValue, errors.EdgeX) {
+	return NewCommandValue(deviceResourceName, v2.ValueTypeBoolArray, value)
```

Comment:
Why do we still need these helper functions? Seems the calling code can easily be switched the use `NewCommandValue`

I assume the calling code is from a Device Service's Driver, but since this is V2 we can have this breaking change that simplifies the code.

Reply:
I don't like helper functions that just call a function without any addition `helper` logic. IMO, these should be removed. Users should understand the valueTypes, otherwise they will have bigger troubles... ;-)

Reply:
removed.